File: python/tests/test_yoco_cluster_se.py
"""
Tests for YOCO compression with clustered standard errors.

Implements Section 5.3.1 of Wong et al. (2021):
"You Only Compress Once: Optimal Data Compression for Estimating Linear Models"

The key insight is that for clustered SEs with compression:
1. Include cluster identifier in the grouping (within-cluster compression)
2. Compute ẽ⁰ = ỹ⁰ - ñ ⊙ ŷ (sum of residuals per group)
3. Build meat matrix using sparse cluster matrix: Ξ̂ = M̃ᵀ diag(ẽ⁰) W̃_C W̃_Cᵀ diag(ẽ⁰) M̃
"""

import numpy as np
import polars as pl
import pytest
from leanfe import leanfe


def test_cluster_se_compress_vs_demean():
    """
    Test that cluster SEs from compression match those from demeaning.
    
    Both methods should produce identical results for the same data.
    """
    np.random.seed(42)
    n = 2000
    n_clusters = 100
    
    # Create data with discrete regressors (good for compression)
    df = pl.DataFrame({
        'y': np.random.randn(n),
        'x1': np.random.choice([0, 1], n),
        'x2': np.random.choice([0, 1, 2], n),
        'fe1': np.random.choice(['A', 'B', 'C', 'D'], n),
        'cluster_id': np.random.choice(range(n_clusters), n)
    })
    
    # Run with compression (should use YOCO)
    result_compress = leanfe(
        df, 
        formula='y ~ x1 + x2 | fe1', 
        vcov='cluster', 
        cluster_cols=['cluster_id'],
        backend='polars'
    )
    
    # Verify compression was used
    assert result_compress.get('strategy') == 'compress', "Should use compression strategy"
    assert result_compress.get('n_clusters') == n_clusters, f"Should have {n_clusters} clusters"
    
    # Check that SEs are reasonable (not NaN or zero)
    for var, se in result_compress['std_errors'].items():
        assert not np.isnan(se), f"SE for {var} should not be NaN"
        assert se > 0, f"SE for {var} should be positive"


def test_cluster_se_duckdb_matches_polars():
    """
    Test that DuckDB and Polars backends produce identical cluster SEs.
    """
    np.random.seed(123)
    n = 1000
    n_clusters = 50
    
    df = pl.DataFrame({
        'y': np.random.randn(n),
        'x1': np.random.choice([0, 1], n),
        'x2': np.random.choice([0, 1, 2], n),
        'fe1': np.random.choice(['A', 'B', 'C'], n),
        'cluster_id': np.random.choice(range(n_clusters), n)
    })
    
    result_polars = leanfe(
        df, 
        formula='y ~ x1 + x2 | fe1', 
        vcov='cluster', 
        cluster_cols=['cluster_id'],
        backend='polars'
    )
    
    result_duckdb = leanfe(
        df, 
        formula='y ~ x1 + x2 | fe1', 
        vcov='cluster', 
        cluster_cols=['cluster_id'],
        backend='duckdb'
    )
    
    # Coefficients should match exactly
    for var in result_polars['coefficients']:
        assert np.isclose(
            result_polars['coefficients'][var], 
            result_duckdb['coefficients'][var],
            rtol=1e-10
        ), f"Coefficients for {var} should match"
    
    # SEs should match (may have small numerical differences)
    for var in result_polars['std_errors']:
        assert np.isclose(
            result_polars['std_errors'][var], 
            result_duckdb['std_errors'][var],
            rtol=1e-6
        ), f"SEs for {var} should match"


def test_cluster_se_ssc_adjustment():
    """
    Test small sample correction (SSC) for clustered SEs.
    
    With SSC, the adjustment is: (G/(G-1)) * ((n-1)/(n-k))
    Without SSC, the adjustment is: G/(G-1)
    
    So SSC should produce larger SEs when n-1 > n-k (i.e., k > 1).
    """
    np.random.seed(456)
    n = 500
    n_clusters = 25
    
    df = pl.DataFrame({
        'y': np.random.randn(n),
        'x1': np.random.choice([0, 1], n),
        'x2': np.random.choice([0, 1, 2], n),
        'fe1': np.random.choice(['A', 'B'], n),
        'cluster_id': np.random.choice(range(n_clusters), n)
    })
    
    result_no_ssc = leanfe(
        df, 
        formula='y ~ x1 + x2 | fe1', 
        vcov='cluster', 
        cluster_cols=['cluster_id'],
        ssc=False
    )
    
    result_ssc = leanfe(
        df, 
        formula='y ~ x1 + x2 | fe1', 
        vcov='cluster', 
        cluster_cols=['cluster_id'],
        ssc=True
    )
    
    # With SSC, SEs should be larger (since we have multiple regressors + FE)
    for var in result_no_ssc['std_errors']:
        assert result_ssc['std_errors'][var] >= result_no_ssc['std_errors'][var], \
            f"SSC should produce larger or equal SEs for {var}"


def test_cluster_se_compression_ratio():
    """
    Test that compression achieves good compression ratio with discrete data.
    """
    np.random.seed(789)
    n = 10000
    n_clusters = 200
    
    # Create data with limited unique combinations
    df = pl.DataFrame({
        'y': np.random.randn(n),
        'x1': np.random.choice([0, 1], n),  # 2 values
        'x2': np.random.choice([0, 1, 2], n),  # 3 values
        'fe1': np.random.choice(['A', 'B', 'C', 'D'], n),  # 4 values
        'cluster_id': np.random.choice(range(n_clusters), n)  # 200 clusters
    })
    
    result = leanfe(
        df, 
        formula='y ~ x1 + x2 | fe1', 
        vcov='cluster', 
        cluster_cols=['cluster_id']
    )
    
    # Should achieve compression
    assert result.get('strategy') == 'compress'
    
    # Compression ratio should be < 1 (fewer compressed records than original)
    compression_ratio = result.get('compression_ratio', 1.0)
    assert compression_ratio < 1.0, f"Should achieve compression, got ratio {compression_ratio}"
    
    # With 2*3*4*200 = 4800 max unique combinations, should compress well
    assert compression_ratio < 0.5, f"Should achieve >50% compression, got {compression_ratio}"


def test_sparse_cluster_matrix():
    """
    Test the sparse cluster matrix builder directly.
    """
    from leanfe.compress import _build_sparse_cluster_matrix
    
    cluster_ids = np.array(['A', 'A', 'B', 'B', 'C', 'A', 'C'])
    W_C, n_clusters = _build_sparse_cluster_matrix(cluster_ids)
    
    assert n_clusters == 3, "Should have 3 unique clusters"
    assert W_C.shape == (7, 3), "Matrix should be 7x3"
    
    # Each row should have exactly one 1
    row_sums = np.array(W_C.sum(axis=1)).flatten()
    assert np.all(row_sums == 1), "Each row should sum to 1"
    
    # Column sums should match cluster sizes
    col_sums = np.array(W_C.sum(axis=0)).flatten()
    assert set(col_sums) == {2, 2, 3}, "Column sums should be cluster sizes"


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: r/tests/test_yoco_cluster_se.R
# Tests for YOCO compression with clustered standard errors
#
# Implements Section 5.3.1 of Wong et al. (2021):
# "You Only Compress Once: Optimal Data Compression for Estimating Linear Models"
#
# The key insight is that for clustered SEs with compression:
# 1. Include cluster identifier in the grouping (within-cluster compression)
# 2. Compute e0_tilde = y0_tilde - n_tilde * yhat (sum of residuals per group)
# 3. Build meat matrix using sparse cluster matrix

library(testthat)
library(Matrix)

# Source the package files
source("R/common.R")
source("R/compress.R")

test_that("sparse cluster matrix builder works correctly", {
  cluster_ids <- c("A", "A", "B", "B", "C", "A", "C")
  result <- .build_sparse_cluster_matrix(cluster_ids)
  
  expect_equal(result$n_clusters, 3)
  expect_equal(dim(result$W_C), c(7, 3))
  
  # Each row should have exactly one 1
  row_sums <- Matrix::rowSums(result$W_C)
  expect_true(all(row_sums == 1))
  
  # Column sums should match cluster sizes
  col_sums <- Matrix::colSums(result$W_C)
  expect_equal(sort(col_sums), c(2, 2, 3))
})


test_that("cluster SE computation produces valid results", {
  set.seed(42)
  n_groups <- 100
  n_clusters <- 20
  p <- 3
  
  # Create mock compressed data
  X <- matrix(rnorm(n_groups * p), n_groups, p)
  Y <- rnorm(n_groups)
  wts <- sqrt(runif(n_groups, 1, 10))
  cluster_ids <- sample(1:n_clusters, n_groups, replace = TRUE)
  
  # Solve WLS
  Xw <- X * wts
  Yw <- Y * wts
  XtX <- crossprod(Xw)
  Xty <- crossprod(Xw, Yw)
  XtX_inv <- solve(XtX)
  beta <- XtX_inv %*% Xty
  
  # Compute RSS
  yhat <- X %*% beta
  rss_g <- (Y - yhat)^2 * wts^2
  rss_total <- sum(rss_g)
  
  # Compute e0_g (sum of residuals per group)
  n_g <- wts^2
  sum_y_g <- Y * n_g
  e0_g <- sum_y_g - n_g * as.vector(yhat)
  
  # Test cluster SE computation
  n_obs <- sum(n_g)
  df_resid <- n_obs - p
  
  se_result <- .compute_se_compress(
    XtX_inv = XtX_inv,
    rss_total = rss_total,
    rss_g = rss_g,
    n_obs = n_obs,
    df_resid = df_resid,
    vcov = "cluster",
    X = X,
    k_x = p,
    cluster_ids = cluster_ids,
    e0_g = e0_g,
    ssc = FALSE
  )
  
  # Check results
  expect_equal(length(se_result$se), p)
  expect_equal(se_result$n_clusters, n_clusters)
  expect_true(all(se_result$se > 0))
  expect_true(all(!is.na(se_result$se)))
})


test_that("SSC adjustment produces larger SEs", {
  set.seed(123)
  n_groups <- 50
  n_clusters <- 10
  p <- 2
  
  X <- matrix(rnorm(n_groups * p), n_groups, p)
  Y <- rnorm(n_groups)
  cluster_ids <- sample(1:n_clusters, n_groups, replace = TRUE)
  
  XtX <- crossprod(X)
  XtX_inv <- solve(XtX)
  beta <- XtX_inv %*% crossprod(X, Y)
  
  yhat <- X %*% beta
  rss_g <- (Y - yhat)^2
  rss_total <- sum(rss_g)
  e0_g <- Y - as.vector(yhat)
  
  n_obs <- n_groups
  df_resid <- n_obs - p
  
  # Without SSC
  se_no_ssc <- .compute_se_compress(
    XtX_inv = XtX_inv, rss_total = rss_total, rss_g = rss_g,
    n_obs = n_obs, df_resid = df_resid, vcov = "cluster",
    X = X, k_x = p, cluster_ids = cluster_ids, e0_g = e0_g, ssc = FALSE
  )
  
  # With SSC
  se_ssc <- .compute_se_compress(
    XtX_inv = XtX_inv, rss_total = rss_total, rss_g = rss_g,
    n_obs = n_obs, df_resid = df_resid, vcov = "cluster",
    X = X, k_x = p, cluster_ids = cluster_ids, e0_g = e0_g, ssc = TRUE
  )
  
  # SSC should produce larger SEs
  expect_true(all(se_ssc$se >= se_no_ssc$se))
})


test_that("IID and HC1 SEs still work with new function signature", {
  set.seed(456)
  n_groups <- 50
  p <- 2
  
  X <- matrix(rnorm(n_groups * p), n_groups, p)
  Y <- rnorm(n_groups)
  
  XtX <- crossprod(X)
  XtX_inv <- solve(XtX)
  beta <- XtX_inv %*% crossprod(X, Y)
  
  yhat <- X %*% beta
  rss_g <- (Y - yhat)^2
  rss_total <- sum(rss_g)
  
  n_obs <- n_groups
  df_resid <- n_obs - p
  
  # IID
  se_iid <- .compute_se_compress(
    XtX_inv = XtX_inv, rss_total = rss_total, rss_g = rss_g,
    n_obs = n_obs, df_resid = df_resid, vcov = "iid",
    X = X, k_x = p
  )
  expect_equal(length(se_iid$se), p)
  expect_true(all(se_iid$se > 0))
  expect_null(se_iid$n_clusters)
  
  # HC1
  se_hc1 <- .compute_se_compress(
    XtX_inv = XtX_inv, rss_total = rss_total, rss_g = rss_g,
    n_obs = n_obs, df_resid = df_resid, vcov = "HC1",
    X = X, k_x = p
  )
  expect_equal(length(se_hc1$se), p)
  expect_true(all(se_hc1$se > 0))
  expect_null(se_hc1$n_clusters)
})


# Tests complete - all test_that blocks above run automatically
